File: src/mbio/packages/denovo_rna/express/express_distribution.r
# __author__ = "qiuping"
# last_modify:20170110
# -*- coding: utf-8 -*-
fpkm = read.table("${input_matrix}", header=TRUE, check.names=0, row.names=1 )
fpkm2 = log(fpkm+1)/log(2)
fpkm10 = log(fpkm+1)/log(10)
fpkm = fpkm
filename = "${filename}"
#fpkm = log(fpkm)/log(2)
#maxfpkm = 20
#samples=colnames(fpkm)
fpkm_density<-function(fpkm_data, logValue){ #add functions by khl 20170403
    samples=colnames(fpkm_data)
    maxfpkm = as.integer(max(fpkm_data)) + 2
    dens=c()
    for(i in 1:length(samples)){
       a=fpkm_data[i]
       den=density(a[,1],n = 601,from = -10,to = maxfpkm)
       dens=cbind(dens,den$y)
    }
    dens_new=cbind(den$x,dens)
    if(logValue == "None"){
       samples = c("fpkm", samples)
    }else{
       samples = c(paste("log", as.character(logValue), "fpkm", sep=""), samples)
    }
    colnames(dens_new)=samples
    return(dens_new)
}
fpkmlog2=fpkm_density(fpkm2,"2")
fpkmlog10=fpkm_density(fpkm10,'10')
fpkm=fpkm_density(fpkm,'None') 
#choose: fpkm>=0.01
fpkmlog2_new = rbind(fpkmlog2[1,], fpkmlog2[-1,][fpkmlog2[-1,1]>=0.1,])
fpkmlog10_new = rbind(fpkmlog10[1,], fpkmlog10[-1,][fpkmlog10[-1,1]>=0.1,])
fpkm_new = rbind(fpkm[1,], fpkm[-1,][fpkm[-1,1]>=0.1,])
write.table(fpkmlog2_new, file=paste("${outputfile}","/log2",filename,"_distribution.xls", sep=""), sep='\t', quote=F, row.names=F, col.names=T)
write.table(fpkmlog10_new, file=paste("${outputfile}","/log10",filename, "_distribution.xls", sep=""), sep='\t', quote=F, row.names=F, col.names=T)
write.table(fpkm_new, file=paste("${outputfile}","/",filename,"_distribution.xls", sep=""), sep='\t', quote=F, row.names=F, col.names=T)
#dens_new<-cbind(den$x,dens)
#samples=c('log2fpkm',samples)
#colnames(dens_new)=samples
#write.table(dens_new, file="/mnt/ilustre/users/sanger-dev/sg-users/konghualei/ref_rna/tofiles/", sep='\t', quote=F, row.names=F, col.names=T)

File: src/mbio/packages/denovo_rna/express/express_distribution.py
#!/mnt/ilustre/users/sanger/app/Python/bin/python
# -*- coding: utf-8 -*-
# __author__ = "qiuping"


from mako.template import Template
from biocluster.config import Config
import os


def distribution(rfile, input_matrix, outputfile, filename):
    """
    生成表达量分布图的数据
    :param input_matrix: 表达量矩阵
    :param outputfile: 输出文件路径
    :param filename: gene或transcript
    """
    this_file_dir = os.path.dirname(os.path.realpath(__file__))
    f = Template(filename=this_file_dir + '/express_distribution.r')
    distribution = f.render(input_matrix=input_matrix, outputfile=outputfile, filename = filename)
    with open(rfile, 'w') as rf:
        rf.write("%s" % distribution)

if __name__ == "__main__":
    distribution("express_distribution.r","/mnt/ilustre/users/sanger-dev/sg-users/konghualei/ref_rna/tofiles/gene_fpkm","/mnt/ilustre/users/sanger-dev/sg-users/konghualei/ref_rna/tofiles/",filename = "gene")
    print 'end'
